### DISCUSSION TITLE
"Build Documentation" Workflow cannot find module `material.extensions`

### DISCUSSION THREAD

**Discussion by mrschick** [squidfunk/mkdocs-material#7523] [Q&A] [answered]
### Context

I am trying to run Materials for MkDocs on a private repo, only issue is that the "Build Documentation" Workflow, which I would like to keep using for future internal PR validation, keeps failing for odd reasons.
At first I thought it was a config issue on my end, but it turns out even unmodified forks of the repo present the issue.

### Bug description

When forking [squidfunk/mkdocs-material](https://github.com/squidfunk/mkdocs-material), the workflow "Build Documentation" will always fail with the following error:
https://github.com/mrschick/mkdocs-material-fork/actions/runs/10848399089/job/30105429423
```
Run mkdocs build --clean
Error: MkDocs encountered an error parsing the configuration file: while constructing a Python object
cannot find module 'material.extensions.emoji' (No module named 'material.extensions')
  in "/home/runner/work/mkdocs-material-fork/mkdocs-material-fork/mkdocs.yml", line 144, column 24
Error: Process completed with exit code 1.
```
When running `mkdocs build --clean` in a local clone of the same forked repo, both with Windows 10 and WSL2 (running Ubuntu-22.04), no errors are thrown and `site` is built correctly.
Removing the affected `markdown_extensions:` from `mkdocs.yaml` only serves to throw a different error on some other extension.

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)

### Reproduction

No minimal reproduction available, issue reproduction requires running Workflows on an unmodified Fork of [squidfunk/mkdocs-material](https://github.com/squidfunk/mkdocs-material).

### Steps to reproduce

1. Make a fork of [squidfunk/mkdocs-material](https://github.com/squidfunk/mkdocs-material);
2. Enable Workflows via the Fork's "Actions" tab;
3. Push an empty commit or a valid documentation change, in order to trigger Workflow execution;
4. The "Build Documentation" Workflow will fail;

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material/), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/) using the [built-in info plugin](https://squidfunk.github.io/mkdocs-material/plugins/info/).

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello @mrschick,
I tested it on my fork and got the same results:
- https://github.com/kamilkrzyskow/mkdocs-material/actions/runs/10849879693/job/30110032554

It turns out that `import material` (and this happens when MkDocs loads the theme) loads the `material` directory as the `material` module. And because of the [sparse-checkout](https://squidfunk.github.io/mkdocs-material/blog/2023/09/22/using-git-sparse-checkout-for-faster-documentation-builds/) the initial checkout doesn't include the full `material` directory and therefore it doesn't work on forks.
https://github.com/squidfunk/mkdocs-material/blob/8df393d11158ebc87026e4eea9ed0c887eae2371/.github/workflows/documentation.yml#L45-L49

It works in the `mkdocs-material` repository, because it fetches the whole Insiders `material` directory and overrides it later.
https://github.com/squidfunk/mkdocs-material/blob/8df393d11158ebc87026e4eea9ed0c887eae2371/.github/workflows/documentation.yml#L89-L90

---

My suggestion would be to use the recommended publishing workflow:
- https://squidfunk.github.io/mkdocs-material/publishing-your-site/

as the workflow from the repository has more of such "hacks" that shouldn't be copied "just because" it's used here.

> [!NOTE]
> In the coming days I'm planning to make a PR with cache handling updates for the publishing guide workflow, so please keep in mind that it will be updated.

**Reply by mrschick** [reply]
Thanks, that finally cleared it up 👌🏻

Since I don't plan on deploying through pages, I didn't consider using that guide's workflow as a base.

**Comment by squidfunk** [comment]
Converting to discussion.
